src/internal/pfsdb/commits.go

```diff
@@ -0,0 +1,956 @@
+		return nil, ErrChildCommitNotFound{ParentRowID: parentCommit}
+	}
+	return children, nil
+}
+
+func getCommitRelativeRows(ctx context.Context, tx *pachsql.Tx, commitID CommitID) (*commitRow, []*commitRow, error) {
+	commitParentRows, err := getCommitParentRow(ctx, tx, commitID)
+	if err != nil && !errors.Is(ErrParentCommitNotFound{ChildRowID: commitID}, errors.Cause(err)) {
+		return nil, nil, errors.Wrap(err, "getting parent commit")
+		// if parent is missing, assume commit is root of a repo.
+	}
+	commitChildrenRows, err := getCommitChildrenRows(ctx, tx, commitID)
+	if err != nil && !errors.Is(ErrChildCommitNotFound{ParentRowID: commitID}, errors.Cause(err)) {
+		return nil, nil, errors.Wrap(err, "getting children commits")
+		// if children is missing, assume commit is HEAD of some branch.
+	}
+	return commitParentRows, commitChildrenRows, nil
+}
+
+func getCommitRowByCommitKey(ctx context.Context, tx *pachsql.Tx, commit *pfs.Commit) (*commitRow, error) {
+	row := &commitRow{}
+	if commit == nil {
+		return nil, ErrCommitMissingInfo{Field: "Commit"}
+	}
+	id := CommitKey(commit)
+	err := tx.QueryRowxContext(ctx, fmt.Sprintf("%s WHERE commit_id=$1", getCommit), id).StructScan(row)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, ErrCommitNotFound{CommitID: id}
+		}
+		return nil, errors.Wrap(err, "scanning commit row")
+	}
+	return row, nil
+}
+
+func parseCommitInfoFromRow(row *commitRow) (*pfs.CommitInfo, error) {
+	commit, err := parseCommitFromRow(row)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting commit from row")
+	}
+	commitInfo := &pfs.CommitInfo{
+		Commit:      commit,
+		Origin:      &pfs.CommitOrigin{Kind: pfs.OriginKind(pfs.OriginKind_value[strings.ToUpper(row.Origin)])},
+		Started:     timeToTimestamppb(row.StartTime),
+		Finishing:   timeToTimestamppb(row.FinishingTime),
+		Finished:    timeToTimestamppb(row.FinishedTime),
+		Description: row.Description,
+		Details: &pfs.CommitInfo_Details{
+			CompactingTime: bigIntToDurationpb(row.CompactingTime),
+			ValidatingTime: bigIntToDurationpb(row.ValidatingTime),
+			SizeBytes:      row.Size,
+		},
+	}
+	return commitInfo, nil
+}
+
+func parseCommitFromRow(row *commitRow) (*pfs.Commit, error) {
+	repo := &pfs.Repo{
+		Name: row.RepoName,
+		Type: row.RepoType,
+		Project: &pfs.Project{
+			Name: row.ProjectName,
+		},
+	}
+	parsedId := strings.Split(row.CommitID, "@")
+	if len(parsedId) != 2 {
+		return nil, errors.New(fmt.Sprintf("got invalid commit id from postgres: %s", row.CommitID))
+	}
+	commit := &pfs.Commit{
+		Repo: repo,
+		Id:   parsedId[1],
+		Branch: &pfs.Branch{
+			Repo: repo,
+			Name: row.BranchName.String,
+		},
+	}
+	return commit, nil
+}
+
+// UpsertCommit will attempt to insert a commit. If the commit already exists, it will update its description.
+func UpsertCommit(ctx context.Context, tx *pachsql.Tx, commitInfo *pfs.CommitInfo, opts ...AncestryOpt) error {
```

Comment:
This is functionally correct, but I would encourage you to remove `CreateCommit` and `UpdateCommit`, and implement both with `UpsertCommit` like how we do it for branches.